File: go/iterator/bookshelf/bookshelf.go
package bookshelf

import (
	"fmt"
)

type Book interface {
	GetName() string
}

type Iterator interface {
	HasNext() bool
	Next() Book
}

type bookShelf struct {
	books   []Book
	last    int
	maxsize int
}

func New(maxsize int) *bookShelf {
	return &bookShelf{
		books:   []Book{},
		last:    0,
		maxsize: maxsize,
	}
}

func (bs *bookShelf) GetBookAt(index int) Book {
	return bs.books[index]
}

func (bs *bookShelf) AppendBook(book Book) error {
	if len(bs.books) >= bs.maxsize {
		return fmt.Errorf("This book shelf had reached the maximum number of books")
	}

	bs.books = append(bs.books, book)
	return nil
}

func (bs *bookShelf) GetLength() int {
	return len(bs.books)
}

func (bs *bookShelf) CreateIterator() Iterator {
	return &shelfIterator{bs, 0}
}

File: typescript/iterator/bookshelf.ts
import { Aggregate } from './interfaces/aggregate.ts';
import { Iterator } from './interfaces/iterator.ts';
import { Book } from './book.ts';

export class BookShelf implements Aggregate {
  private books: Book[];

  constructor() {
    this.books = [];
  }

  appendBook = (book: Book): void => {
    this.books.push(book);
  };

  getBookAt = (index: number): Book => {
    return this.books[index];
  };

  getLength = (): number => {
    return this.books.length;
  };

  createIterator = (): Iterator => {
    return new ShelfIterator(this);
  }
}

class ShelfIterator implements Iterator {
  private bookShelf: BookShelf;
  private index: number;

  constructor(bookShelf: BookShelf) {
    this.bookShelf = bookShelf;
    this.index = 0;
  }

  hasNext = (): boolean => {
    if (this.index < this.bookShelf.getLength()) {
      return true;
    }
    return false;
  };

  next = (): Book | null => {
    if (!this.hasNext()) {
      return null;
    }
    const nextBook = this.bookShelf.getBookAt(this.index);
    this.index++;
    return nextBook;
  }
}